Mandd/et import analysis

mandd opened the pull request:
--------
Pull Request Description
--------
##### What issue does this change request address? (Use "#" before the issue to link it, i.e., #42.)
Closes #422 

##### What are the significant changes in functionality due to this change request?


----------------
For Change Control Board: Change Request Review
----------------
The following review must be completed by an authorized member of the Change Control Board.
- [x] 1. Review all computer code. __yes__
- [x] 2. If any changes occur to the input syntax, there must be an accompanying change to the user manual and xsd schema. If the input syntax change deprecates existing input files, a conversion script needs to be added (see Conversion Scripts). __this is a new feature, user manual is updated__
- [x] 3. Make sure the Python code and commenting standards are respected (camelBack, etc.) - See on the [wiki](https://github.com/idaholab/raven/wiki/RAVEN-Code-Standards#python) for details. __yes__
- [x] 4. Automated Tests should pass, including run_tests, pylint, manual building and xsd tests. If there are changes to Simulation.py or JobHandler.py the qsub tests must pass. __all pass__
- [x] 5. If significant functionality is added, there must  be tests added to check this. Tests should cover all possible options.  Multiple short tests are preferred over one large test. If new development on the internal JobHandler parallel system is performed, a cluster test must be added setting, in <RunInfo> XML block, the node ```<internalParallel>``` to True. __added__
- [x] 6. If the change modifies or adds a requirement or a requirement based test case, the Change Control Board's Chair or designee also needs to approve the change.  The requirements and the requirements test shall be in sync. __N/A__
- [x] 7. The merge request must reference an issue.  If the issue is closed, the issue close checklist shall be done. __yes__
- [x] 8. If an analytic test is changed/added is the the analytic documentation updated/added? __N/A__

Checklists reviewed by Congjian Wang @wangcj05 on Nov. 22, 2017

alfoa commented on framework/PostProcessors/ETimporter.py:
```diff
@@ -0,0 +1,262 @@
+        specifying input of cls.
+    """
+    inputSpecification = super(ETimporter, cls).getInputSpecification()
+
```
you can remove this line

mandd replied:
done

alfoa commented on framework/PostProcessors/ETimporter.py:
```diff
@@ -0,0 +1,262 @@
+    """
+    PostProcessor.initialize(self, runInfo, inputs, initDict)
+    self.inputs = inputs
+    self._workingDir = runInfo['WorkingDir']
```
do we really need to do that? Isn't the PostProcessor base class already storing these variables?

wangcj05 replied:
I think you can remove this line, this variable is now available in the PP base class.

alfoa commented on framework/PostProcessors/ETimporter.py:
```diff
@@ -0,0 +1,262 @@
+      @ In,  input, object, object containing the data to process. (inputToInternal output)
+      @ Out, None
+    """
+    if self.ETformat == 'OpenPSA':
```
this if statement is probably redundant with the current implementation since the OpenPSA is the only one we support (and you check for it before).
If you want to keep it, I would use a lambda function with a dictionary mapping

alfoa commented on framework/PostProcessors/ETimporter.py:
```diff
@@ -0,0 +1,262 @@
+  def runOpenPSA(self, input):
+    """
+      This method executes the postprocessor action.
+      @ In,  input, object, object containing the data to process. (inputToInternal output)
```
which object?

alfoa commented on framework/PostProcessors/ETimporter.py:
```diff
@@ -0,0 +1,262 @@
+  def run(self, input):
+    """
+      This method executes the postprocessor action.
+      @ In,  input, object, object containing the data to process. (inputToInternal output)
```
which object?

alfoa commented on framework/PostProcessors/ETimporter.py:
```diff
@@ -0,0 +1,262 @@
+      self.raiseAWarning('Output type ' + str(output.type) + ' is not supported.')
+
+  def findAllRecursive(self, node, element, result = None):
+    """
```
I think the same outcome is achievable with the ```Element.findall()``` method

mandd replied:
this method differs from findall() since it goes for all nodes, subnodes, subsubnodes etc. recursively

alfoa replied:
Sorry my bad... It is not find all but ```iter```
For example, 
```
import xml.etree.ElementTree as ET
results = node.iter(element)
```

mandd replied:
done!

alfoa commented on framework/PostProcessors/ETimporter.py:
```diff
@@ -0,0 +1,262 @@
+    if isinstance(evaluation, Runners.Error):
+      self.raiseAnError(RuntimeError, ' No available output to collect (Run probably is not finished yet) via',self.printTag)
+    if not set(output.getParaKeys('inputs')) == set(self.variables):
+      self.raiseAnError(RuntimeError, ' ETimporter: set of branching variables in the ET ( ' + str(output.getParaKeys('inputs')) + ' ) is not identical to the set of input variables specified in the PointSet (' + str(self.variables) +')')
```
can you split the line in two lines?

wangcj05 commented on doc/user_manual/postprocessor.tex:
```diff
@@ -1617,6 +1616,56 @@ \subsubsection{RavenOutput}
 
+%%%%%%%%%%%%%% ETimporter PP %%%%%%%%%%%%%%%%%%%
+
+\subsubsection{ETimporter}
```
Should we use ```ETImporter```?

maljovec replied:
If you are referring to the case change, then I would agree with @wangcj05

wangcj05 commented on doc/user_manual/postprocessor.tex:
```diff
@@ -1617,6 +1616,56 @@ \subsubsection{RavenOutput}
+
+\subsubsection{ETimporter}
+\label{ETimporterPP}
+The \textbf{ETimporter} post-processor has been designed to import Event-Tree (ETs) object into
```
change ```ETs``` to ```ET```

wangcj05 commented on framework/PostProcessors/ETimporter.py:
```diff
@@ -0,0 +1,262 @@
+#Internal Modules End-----------------------------------------------------------
+
+
+class ETimporter(PostProcessor):
```
Same as before, should we change the class name to ```ETImporter``` for consistence?

wangcj05 commented on framework/PostProcessors/ETimporter.py:
```diff
@@ -0,0 +1,262 @@
+
+class ETimporter(PostProcessor):
+  """
+    This is the base class for postprocessors
```
The description is not related, please change it.

wangcj05 commented on framework/PostProcessors/ETimporter.py:
```diff
@@ -0,0 +1,262 @@
+    """
+    PostProcessor.__init__(self, messageHandler)
+    self.printTag = 'POSTPROCESSOR ET IMPORTER'
+    self.ETformat = None
```
change ```ETformat``` to ```ETFormat```

wangcj05 commented on framework/PostProcessors/ETimporter.py:
```diff
@@ -0,0 +1,262 @@
+
+
+  @classmethod
+  def getInputSpecification(cls):
```
I think we should fill in this method, this is good for input validation.

mandd replied:
added

wangcj05 commented on framework/PostProcessors/ETimporter.py:
```diff
@@ -0,0 +1,262 @@
+      @ In, xmlNode, xml.etree.Element, Xml element node
+      @ Out, None
+    """
+    for child in xmlNode:
```
As mentioned before, we should start to fill in the getInputSpecification method, and using InputData for this method. Please let me know if you have any questions.

mandd replied:
done!

wangcj05 commented on framework/PostProcessors/ETimporter.py:
```diff
@@ -0,0 +1,262 @@
+      else:
+        self.raiseAnError(IOError, 'ETimporterPostProcessor Post-Processor ' + self.name + ', node ' + child.tag + ' : is not recognized')
+
+  def run(self, input):
```
I think we should use different name for ```input``` here, since ```input``` is a reserve function in python

wangcj05 commented on framework/PostProcessors/ETimporter.py:
```diff
@@ -0,0 +1,262 @@
+    if self.ETformat == 'OpenPSA':
+      self.outputDict = self.runOpenPSA(input)
+
+  def runOpenPSA(self, input):
```
same here as before

wangcj05 commented on framework/PostProcessors/ETimporter.py:
```diff
@@ -0,0 +1,262 @@
+    outcomes = []
+
+    ## These self.variables will be mapped into an array where there index
+    self.variables = []
```
please do not use the class variables in the run method if the class variables are changing during the run. This will create issues when you run in parallel. See https://github.com/idaholab/raven/wiki/Avoid-Using-Model-State-when-in-Parallel-Execution-Context

wangcj05 commented on framework/PostProcessors/ETimporter.py:
```diff
@@ -0,0 +1,262 @@
+
+      ## Iterate through the forks that use this event and gather all of the
+      ## possible states
+      for fork in self.findAllRecursive(eventTree, 'fork'):
```
I think you can use ```Element.iter()``` method from python ET instead to create your own method here.

mandd commented:
I have added Dan to the reviewer; he did most of the work contained in this PR

wangcj05 commented on framework/PostProcessors/ETimporter.py:
```diff
@@ -0,0 +1,262 @@
+    d = len(self.variables)
+    n = len(self.findAllRecursive(eventTree, 'sequence'))
+
+    self.pointSet = -1 * np.ones((n, d + 1))
```
Same as before, do not use the state variables if they will change during the run time.

maljovec commented on framework/PostProcessors/ETimporter.py:
```diff
@@ -0,0 +1,262 @@
+    eventTree = root.findall('initial-state')
+
+    if len(eventTree) > 1:
+      self.raiseAnError(IOError,'ETimporter: more than one initial-state identified')
```
`ETImporter`, also if you wanted you could make this more dynamic: `self.__class__.__name__` and then if you renamed this class or had a subclass that called this method, then this would automatically use the correct identifier.

maljovec commented on framework/PostProcessors/Factory.py:
```diff
@@ -51,6 +51,7 @@
   pass
 
 ## [ Add new class here ]
+from .ETimporter import ETimporter
```
`ETimporter` -> `ETImporter`

maljovec commented on framework/PostProcessors/__init__.py:
```diff
@@ -70,4 +70,5 @@
            'Metric',
-           'CrossValidation'] + additionalModules
+           'CrossValidation',
+           'ETimporter'] + additionalModules
```
`ETimporter` -> `ETImporter`

mandd commented:
By talking with Conjiang and Andrea, this PP does not handle certain OpenPSA ET features such as 
<define-branch>
and:
<define-sequence name="Link-to-ET-A">
    <event-tree name="ET-A"/>
</define-sequence> 
I need a bit more time to address this before addressing the comments

alfoa commented:
Ok.

Let us know when you are ready for the second review.

mandd commented:
I extended the capabilities and addresses the comments of the reviewers.
I suggest a second round of review since a lot of things has changed

moosebuild commented:
Job [Mingw Test](https://www.moosebuild.org/job/135983/) on 230621c : invalidated by @mandd

wangcj05 commented on doc/user_manual/postprocessor.tex:
```diff
@@ -1620,6 +1619,78 @@ \subsubsection{RavenOutput}
+Fault-Trees to the branch (i.e., fork) point.
+
+%
+\ppType{ETimporter}{ETimporter}
```
changed to ```ETImporter```

wangcj05 commented on doc/user_manual/postprocessor.tex:
```diff
@@ -1620,6 +1619,78 @@ \subsubsection{RavenOutput}
+ ...
+  <Models>
+    ...
+    <PostProcessor name="ETimporter" subType="ETimporter">
```
subType should be changed to ETImporter

wangcj05 commented on framework/PostProcessors/ETimporter.py:
```diff
@@ -0,0 +1,554 @@
+# Copyright 2017 Battelle Energy Alliance, LLC
```
Rename the file name ETimporter.py to ETImporter.py

wangcj05 commented on framework/PostProcessors/ETimporter.py:
```diff
@@ -0,0 +1,554 @@
+
+    return inputSpecification
+
+
```
remove extra empty line

wangcj05 commented on framework/PostProcessors/ETimporter.py:
```diff
@@ -0,0 +1,554 @@
+    paramInput.parseNode(xmlNode)
+    self._handleInput(paramInput)
+
+  def _handleInput(self, paramInput):
```
add doc string

wangcj05 commented on framework/PostProcessors/ETimporter.py:
```diff
@@ -0,0 +1,554 @@
+      @ In,  inputs, list, list of file objects
+      @ Out, None
+    """
+    self.outputDict = self.runOpenPSA(inputs)
```
I think we should avoid changing the state variable self.outputDict, because this not thread safe. Please use local variable. see https://github.com/idaholab/raven/wiki/Avoid-Using-Model-State-when-in-Parallel-Execution-Context

wangcj05 commented on framework/PostProcessors/ETimporter.py:
```diff
@@ -0,0 +1,554 @@
+    """
+    self.outputDict = self.runOpenPSA(inputs)
+
+  def runOpenPSA(self, inputs):
```
same as before, in this method, we should avoid changing the state variables. such as self.links, self. connectivityMatrix, self.listETs, etc.

wangcj05 commented on framework/PostProcessors/ETimporter.py:
```diff
@@ -0,0 +1,554 @@
+    self.listRoots=[]
+
+    for file in inputs:
+        EventTree = ET.parse(file.getPath() + file.getFilename())
```
should be ```eventTree```?

wangcj05 commented on framework/PostProcessors/ETimporter.py:
```diff
@@ -0,0 +1,554 @@
+        self.raiseAnError(IOError, 'A single ET files has provided but it contains a link to an additional ET')
+
+    if len(self.links)==0 and len(inputs)==1:
+        EventTree = ET.parse(inputs[0].getPath() + inputs[0].getFilename())
```
should be eventTree?

wangcj05 commented on framework/PostProcessors/ETimporter.py:
```diff
@@ -0,0 +1,554 @@
+
+    if len(self.links)==0 and len(inputs)==1:
+        EventTree = ET.parse(inputs[0].getPath() + inputs[0].getFilename())
+        return self.analyzeSingleET(EventTree.getroot())
```
same as before

wangcj05 commented on framework/PostProcessors/ETimporter.py:
```diff
@@ -0,0 +1,554 @@
+        EventTree = ET.parse(inputs[0].getPath() + inputs[0].getFilename())
+        return self.analyzeSingleET(EventTree.getroot())
+
+  def createLinkList(self):
```
should avoid changing the state variables, such as self.links

wangcj05 commented on framework/PostProcessors/ETimporter.py:
```diff
@@ -0,0 +1,554 @@
+                  self.links.append(dep)
+
+
+  def checkETstructure(self):
```
see previous comments

wangcj05 commented on framework/PostProcessors/ETimporter.py:
```diff
@@ -0,0 +1,554 @@
+      if len(zeroColumns)==0:
+          self.raiseAnError(IOError, 'No root ET')
+      if len(zeroColumns)==1:
+          self.rootET_ID = self.listETs[zeroColumns]
```
underscore ```_``` should be avoided

wangcj05 commented on framework/PostProcessors/ETimporter.py:
```diff
@@ -0,0 +1,554 @@
+      self.raiseADebug("ETImporter leaf ETs: " + str(leafs))
+
+
+  def analyzeMultipleET(self,input):
```
I suggest to use a different name for input, because input is a function in python

wangcj05 replied:
state variable should be avoided in this method, such as self.listET, ..., please use local variables

wangcj05 commented on framework/PostProcessors/ETimporter.py:
```diff
@@ -0,0 +1,554 @@
+      return self.listRoots[indexRootET]
+
+
+  def analyzeSingleET(self,masterRoot):
```
see previous comments.

mandd replied:
done

wangcj05 commented on framework/PostProcessors/ETimporter.py:
```diff
@@ -0,0 +1,554 @@
+
+      return outputDict
+
+
```
remove extra empty line

wangcj05 commented on framework/PostProcessors/ETimporter.py:
```diff
@@ -0,0 +1,554 @@
+        The name of the <event-tree> is the link to the ET defined as follows:
+            <define-event-tree name="Link-to-LP-Event-Tree">
+
+        @ In,  masterRoot, xml.etree.Element, root of the root ET
```
root, instead of masterRoot.

wangcj05 commented on framework/PostProcessors/ETimporter.py:
```diff
@@ -0,0 +1,554 @@
+                      node.remove(subNode)
+      return copy.deepcopy(rootMaster)
+
+
```
remove extra empty line

wangcj05 commented on framework/PostProcessors/ETimporter.py:
```diff
@@ -0,0 +1,554 @@
+              map[seq] = float(seq)
+      return map
+
+
```
remove extra empty line

wangcj05 commented on framework/PostProcessors/ETimporter.py:
```diff
@@ -0,0 +1,554 @@
+        for value in self.outputDict['outputs'][key]:
+          output.updateOutputValue(str(key),value)
+    else:
+      self.raiseAWarning('Output type ' + str(output.type) + ' is not supported.')
```
I think an error should be raised here.

alfoa replied:
agree

wangcj05 commented on framework/PostProcessors/ETimporter.py:
```diff
@@ -0,0 +1,554 @@
+
+
+
+
```
remove empty lines

wangcj05 commented on framework/PostProcessors/Factory.py:
```diff
@@ -51,6 +51,7 @@
   pass
 
 ## [ Add new class here ]
+from .ETimporter import ETImporter
```
change ETimporter to ETImporter

alfoa commented on framework/PostProcessors/ETimporter.py:
```diff
@@ -0,0 +1,554 @@
+          outcome = node.get('name')
+          if outcome not in outcomes:
+              outcomes.append(outcome)
+      map = self.returnMap(outcomes, root.get('name'))
```
map is a intrinsic python function...the variable name should be changed

alfoa commented on framework/PostProcessors/ETimporter.py:
```diff
@@ -0,0 +1,554 @@
+      allFloat = True
+      for seq in outcomes:
+          try:
+              float(seq)
```
we have utility functions for achieving this... check mathUtils.py and utils.py

mandd replied:
done!

alfoa commented on framework/PostProcessors/ETimporter.py:
```diff
@@ -0,0 +1,554 @@
+          except ValueError:
+              allFloat = False
+              break
+      map = {}
```
map is an intrinsic python function

alfoa commented on framework/PostProcessors/ETimporter.py:
```diff
@@ -0,0 +1,554 @@
+          updatedTreeMap.write(fileID)
+      else:
+          for seq in outcomes:
+              map[seq] = float(seq)
```
utility functiokn

mandd replied:
edited

alfoa commented on framework/PostProcessors/ETimporter.py:
```diff
@@ -0,0 +1,554 @@
+    """
+    evaluation = finishedJob.getEvaluation()
+    if isinstance(evaluation, Runners.Error):
+      self.raiseAnError(RuntimeError, ' No available output to collect (Run probably is not finished yet) via',self.printTag)
```
I would change the message... Run failed

alfoa commented on tests/framework/PostProcessors/ETimporterPostProcessor/gold/ETimporter/PrintPS.csv:
```diff
@@ -0,0 +1,5 @@
+ACC,LPR,LPI,sequence
```
it is not for this file but
The file ```tests/framework/PostProcessors/ETimporterPostProcessor/gold/.DS_Store```
should be removed.
(see above)

alfoa commented on tests/framework/PostProcessors/ETimporterPostProcessor/test_ETimporter.xml:
```diff
@@ -0,0 +1,54 @@
+    <created>2017-11-07</created>
+    <classesTested>ETimporter</classesTested>
+    <description>
+      Tests of the ETimporter post-processor
```
a bit of more description here.

alfoa commented on tests/framework/PostProcessors/ETimporterPostProcessor/test_ETimporterMultipleET.xml:
```diff
@@ -0,0 +1,56 @@
+    <created>2017-11-07</created>
+    <classesTested>ETimporter</classesTested>
+    <description>
+      Tests of the ETimporter post-processor with multiple ETs
```
more description

PS. classesTested should be ```Models.PostProcessors.ETImporter```

In addition, the name should be equal to how to run using the run_test command
In this case from 
```framework/PostProcessors/ETimporterPostProcessor```
it should be
```framework/PostProcessors.ETimporterPostProcessor```

since we can run this witht he following command

```./run_tests --re=framework/PostProcessors.ETimporterPostProcessor```

alfoa commented on tests/framework/PostProcessors/ETimporterPostProcessor/test_ETimporterSymbolic.xml:
```diff
@@ -0,0 +1,54 @@
+<Simulation verbosity="debug">
+  <TestInfo>
+    <name>framework/PostProcessors/ETimporterPostProcessor</name>
```
same as above.

In addition all the <name> nodes you create in the tests have the same name...

Remeber that this names are used by the automatic regression test documentation latex script

alfoa commented on tests/framework/PostProcessors/ETimporterPostProcessor/test_ETimporter_DefineBranch.xml:
```diff
@@ -0,0 +1,54 @@
+    <created>2017-11-07</created>
+    <classesTested>ETimporter</classesTested>
+    <description>
+      Tests of the ETimporter post-processor with branch defined
```
same as above

alfoa commented:
@mandd I did my review... In addition there are a lot of ```.DS_Store``` files that should be removed.

mandd commented:
I have completed the addressing of this second round of review

wangcj05 commented on framework/PostProcessors/ETimporter.py:
```diff
@@ -0,0 +1,530 @@
+  def analyzeMultipleET(self,inputs,links,listRoots,listETs,rootETID):
+    """
+      This method executes the analysis of the ET if multiple ETs are provided. It merge all ETs onto the root ET
+      @ In,  input, list, list of file objects
```
update doc string here

wangcj05 commented on framework/PostProcessors/ETimporter.py:
```diff
@@ -0,0 +1,530 @@
+    # 1. for all ET check if it contains SubBranches
+    ETset = []
+    for fileID in inputs:
+      EventTree = ET.parse(fileID.getPath() + fileID.getFilename())
```
should be eventTree?

wangcj05 commented on framework/PostProcessors/ETimporter.py:
```diff
@@ -0,0 +1,530 @@
+    ## These outcomes will be encoded as integers starting at 0
+    outcomes = []
+    ## These self.variables will be mapped into an array where there index
+    self.variables = []
```
state variable should be avoided.

wangcj05 commented on framework/PostProcessors/ETimporter.py:
```diff
@@ -0,0 +1,530 @@
+
+    d = len(self.variables)
+    n = len(self.findAllRecursive(root.find('initial-state'), 'sequence'))
+    self.pointSet = -1 * np.ones((n, d + 1))
```
state variable should be avoided

mandd commented:
file name has been updated: checking tests

mandd commented:
The last comments from @wangcj05 were removed from github but I have addressed them as follows:
@wangcj05 requested changes on this pull request.

I have some comments, in addition, I suggest rename the file framework/PostProcessors/ETimporter.py to framework/PostProcessors/ETImporter.py

In framework/PostProcessors/ETimporter.py:

> +      @ In, paramInput, ParameterInput, the already parsed input.
+      @ Out, None
+    """
+    fileFormat = paramInput.findFirst('fileFormat')
+    self.fileFormat = fileFormat.value
+    if self.fileFormat not in self.allowedFormats:
+      self.raiseAnError(IOError,
+                          'ETImporterPostProcessor Post-Processor ' + self.name + ', format ' + str(self.fileFormat) + ' : is not supported')
+
+  def run(self, inputs):
+    """
+      This method executes the postprocessor action.
+      @ In,  inputs, list, list of file objects
+      @ Out, None
+    """
+    #self.outputDict = self.runOpenPSA(inputs)
remove the commented line
@mandd: DONE
In framework/PostProcessors/ETimporter.py:

> +      dictionary as follows:
+
+          dep.keys =
+                     * link_seqID  : ID of the link in the master ET
+                     * ET_slave_ID : ID of the slave ET that needs to be copied into the master ET
+                     * ET_master_ID: ID of the master ET;
+
+      The slave ET is merged into the master ET; note the master ET contains the link in at least one
+      <define-sequence> node:
+
+         <define-sequence name="Link-to-LP">
+           <event-tree name="Link-to-LP-Event-Tree"/>
+         </define-sequence>
+
+      @ In,  None
+      @ Out, None
update the doc string here
@mandd: DONE
In framework/PostProcessors/ETimporter.py:

> +          dep = {}
+          dep['link_seqID'] = copy.deepcopy(seqID[idx])
+          dep['ET_slave_ID'] = copy.deepcopy(val)
+          dep['ET_master_ID'] = copy.deepcopy(root.get('name'))
+          linkList.append(dep)
+    return linkList
+
+  def checkETstructure(self,links,listETs,connectivityMatrix):
+    """
+      This method checks that the structure of the ET is consistent. In particular, it checks that only one root ET
+      and at least one leaf ET is provided. As an example consider the following ET structure:
+         ET1 ----> ET2 ----> ET3
+          |------> ET4 ----> ET5
+      Five ETs have been provided, ET1 is the only root ET while ET3 and ET5 are leaf ET.
+      @ In,  None
+      @ Out, None
update doc string here
@mandd: DONE
In framework/PostProcessors/ETimporter.py:

> +    if len(zeroColumns)>1:
+      self.raiseAnError(IOError, 'Multiple root ET')
+    if len(zeroColumns)==0:
+      self.raiseAnError(IOError, 'No root ET')
+    if len(zeroColumns)==1:
+      rootETID = listETs[zeroColumns]
+      self.raiseADebug("ETImporter Root ET: " + str(rootETID))
+
+    leafs = []
+    for index in np.nditer(zeroRows):
+      leafs.append(listETs[index])
+    self.raiseADebug("ETImporter leaf ETs: " + str(leafs))
+
+    return rootETID
+
+
remove extra empty line
@mandd: DONE
In framework/PostProcessors/ETimporter.py:

> +    if len(zeroColumns)==1:
+      rootETID = listETs[zeroColumns]
+      self.raiseADebug("ETImporter Root ET: " + str(rootETID))
+
+    leafs = []
+    for index in np.nditer(zeroRows):
+      leafs.append(listETs[index])
+    self.raiseADebug("ETImporter leaf ETs: " + str(leafs))
+
+    return rootETID
+
+
+  def analyzeMultipleET(self,inputs,links,listRoots,listETs,rootETID):
+    """
+      This method executes the analysis of the ET if multiple ETs are provided. It merge all ETs onto the root ET
+      @ In,  input, list, list of file objects
update doc string here
@mandd: DONE
In framework/PostProcessors/ETimporter.py:

> +      leafs.append(listETs[index])
+    self.raiseADebug("ETImporter leaf ETs: " + str(leafs))
+
+    return rootETID
+
+
+  def analyzeMultipleET(self,inputs,links,listRoots,listETs,rootETID):
+    """
+      This method executes the analysis of the ET if multiple ETs are provided. It merge all ETs onto the root ET
+      @ In,  input, list, list of file objects
+      @ Out, xmlNode, xml.etree.Element, root of the assembled root ET
+    """
+    # 1. for all ET check if it contains SubBranches
+    ETset = []
+    for fileID in inputs:
+      EventTree = ET.parse(fileID.getPath() + fileID.getFilename())
should be eventTree?
@mandd: DONE
In framework/PostProcessors/ETimporter.py:

> +
+    indexRootET = listETs.index(rootETID)
+    return listRoots[indexRootET]
+
+  def analyzeSingleET(self,masterRoot):
+    """
+      This method executes the analysis of the ET if a single ET is provided.
+      @ In,  masterRoot, xml.etree.Element, root of the ET
+      @ Out, outputDict, dict, dictionary containing the pointSet data
+    """
+    root = self.checkSubBranches(masterRoot)
+
+    ## These outcomes will be encoded as integers starting at 0
+    outcomes = []
+    ## These self.variables will be mapped into an array where there index
+    self.variables = []
state variable should be avoided.

In framework/PostProcessors/ETimporter.py:

> +            if state not in values[event]:
+              values[event].append(state)
+
+    ## Iterate through the sequences and gather all of the possible outcomes
+    ## so we can numerically encode them latter
+    for node in root.findall('define-sequence'):
+      outcome = node.get('name')
+      if outcome not in outcomes:
+        outcomes.append(outcome)
+    etMap = self.returnMap(outcomes, root.get('name'))
+
+    self.raiseADebug("ETImporter variables identified: " + str(format(self.variables)))
+
+    d = len(self.variables)
+    n = len(self.findAllRecursive(root.find('initial-state'), 'sequence'))
+    self.pointSet = -1 * np.ones((n, d + 1))
state variable should be avoided

wangcj05 commented:
@mandd by the way, you can still see my comments, just click "Show outdated"

wangcj05 commented:
When all tests are green, I will do the checklists. Great Job!

wangcj05 commented:
Everything looks good now. The checklists are satisfied. This PR is good to go.

In addition, an email is preferred to the users for this new feature. @mandd do we have a place to document this feature in order to accelerate the new release process?

wangcj05 commented:
Happy thanksgiving, and enjoy the holiday. Merging
